File: Linked List/1171.remove-zero-sum-consecutive-nodes-from-linked-list.java
/*
 * @lc app=leetcode id=1171 lang=java
 *
 * [1171] Remove Zero Sum Consecutive Nodes from Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        Map<Integer, ListNode> map = new HashMap<>();
        
        // 首次遍历建立哈希表，若同一sum出现多次仅记录该sum最后一次出现节点
        int sum = 0;
        for(ListNode d = dummy; d != null; d = d.next) {
            sum += d.val;
            map.put(sum, d);
        }

        // 第二次遍历，若当前节点对应的sum在后面的节点可找到对应，表明两节点中间所有节点和为0
        sum = 0;
        for(ListNode d = dummy; d != null; d = d.next) {
            sum += d.val;
            d.next = map.get(sum).next;
        }
        return dummy.next;
    }
}
// @lc code=end

File: Linked List/1171.remove-zero-sum-consecutive-nodes-from-linked-list.py
#
# @lc app=leetcode id=1171 lang=python3
#
# [1171] Remove Zero Sum Consecutive Nodes from Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeZeroSumSublists(self, head: ListNode) -> ListNode:
        dummy = ListNode(0)
        dummy.sum = 0
        dummy.next = head
        cur_sum = 0
        p = head
        while p:
            cur_sum += p.val
            p.sum = cur_sum
            p = p.next
        start = dummy
        while start:
            cur = start.next
            while cur:
                range_sum = cur.sum - start.sum
                if range_sum == 0:
                    start.next = cur.next
                cur = cur.next
            start = start.next
        return dummy.next
# @lc code=end

